##COVID-19 Analysis
This notebook processes and performs quick analysis from the [New York Times COVID-19 dataset](https://github.com/nytimes/covid-19-data).  The data is updated in the `/databricks-datasets/COVID/covid-19-data/` location regularly so you can access the data directly.

In [2]:
%sh pip install altair

Collecting altair
 Downloading altair-4.1.0-py3-none-any.whl (727 kB)
Collecting jinja2
 Downloading Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
Requirement already satisfied: entrypoints in /databricks/python3/lib/python3.7/site-packages (from altair) (0.3)
Collecting toolz
 Downloading toolz-0.10.0.tar.gz (49 kB)
Requirement already satisfied: pandas>=0.18 in /databricks/python3/lib/python3.7/site-packages (from altair) (1.0.1)
Collecting jsonschema
 Downloading jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Requirement already satisfied: numpy in /databricks/python3/lib/python3.7/site-packages (from altair) (1.18.1)
Collecting MarkupSafe>=0.23
 Downloading MarkupSafe-1.1.1-cp37-cp37m-manylinux1_x86_64.whl (27 kB)
Requirement already satisfied: python-dateutil>=2.6.1 in /databricks/python3/lib/python3.7/site-packages (from pandas>=0.18->altair) (2.8.1)
Requirement already satisfied: pytz>=2017.2 in /databricks/python3/lib/python3.7/site-packages (from pandas>=0.18->altair) (2019.3)
Collecting importlib-metadata; python_version < "3.8"
 Downloading importlib_metadata-1.7.0-py2.py3-none-any.whl (31 kB)
Requirement already satisfied: six>=1.11.0 in /databricks/python3/lib/python3.7/site-packages (from jsonschema->altair) (1.14.0)
Collecting attrs>=17.4.0
 Downloading attrs-19.3.0-py2.py3-none-any.whl (39 kB)
Collecting pyrsistent>=0.14.0
 Downloading pyrsistent-0.16.0.tar.gz (108 kB)
Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from jsonschema->altair) (45.2.0)
Collecting zipp>=0.5
 Downloading zipp-3.1.0-py3-none-any.whl (4.9 kB)
Building wheels for collected packages: toolz, pyrsistent
 Building wheel for toolz (setup.py): started
 Building wheel for toolz (setup.py): finished with status 'done'
 Created wheel for toolz: filename=toolz-0.10.0-py3-none-any.whl size=55575 sha256=1f9e1c2dc143f8b2c4cf28d2b109a25627be1e93abbd21ef66d8513cd0cade96
 Stored in directory: /root/.cache/pip/wheels/e2/83/7c/248063997a4f9ff6bf145822e620e8c37117a6b4c765584077
 Building wheel for pyrsistent (setup.py): started
 Building wheel for pyrsistent (setup.py): finished with status 'done'
 Created wheel for pyrsistent: filename=pyrsistent-0.16.0-cp37-cp37m-linux_x86_64.whl size=98728 sha256=b66e917522656195bd3d903e372bc51ff8a49671d8c3b38ec6d93a63717cad72
 Stored in directory: /root/.cache/pip/wheels/22/52/11/f0920f95c23ed7d2d0b05f2b7b2f4509e87a20cfe8ea43d987
Successfully built toolz pyrsistent
Installing collected packages: MarkupSafe, jinja2, toolz, zipp, importlib-metadata, attrs, pyrsistent, jsonschema, altair
Successfully installed MarkupSafe-1.1.1 altair-4.1.0 attrs-19.3.0 importlib-metadata-1.7.0 jinja2-2.11.2 jsonschema-3.2.0 pyrsistent-0.16.0 toolz-0.10.0 zipp-3.1.0
WARNING: You are using pip version 20.0.2; however, version 20.1.1 is available.
You should consider upgrading via the '/databricks/python3/bin/python3.7 -m pip install --upgrade pip' command.

In [3]:
%sh pip install vega_datasets

Collecting vega_datasets
 Downloading vega_datasets-0.8.0-py2.py3-none-any.whl (210 kB)
Requirement already satisfied: pandas in /databricks/python3/lib/python3.7/site-packages (from vega_datasets) (1.0.1)
Requirement already satisfied: python-dateutil>=2.6.1 in /databricks/python3/lib/python3.7/site-packages (from pandas->vega_datasets) (2.8.1)
Requirement already satisfied: numpy>=1.13.3 in /databricks/python3/lib/python3.7/site-packages (from pandas->vega_datasets) (1.18.1)
Requirement already satisfied: pytz>=2017.2 in /databricks/python3/lib/python3.7/site-packages (from pandas->vega_datasets) (2019.3)
Requirement already satisfied: six>=1.5 in /databricks/python3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->vega_datasets) (1.14.0)
Installing collected packages: vega-datasets
Successfully installed vega-datasets-0.8.0
WARNING: You are using pip version 20.0.2; however, version 20.1.1 is available.
You should consider upgrading via the '/databricks/python3/bin/python3.7 -m pip install --upgrade pip' command.

In [4]:
# Standard Libraries
import io

# External Libraries
import requests
import numpy as np
import pandas as pd
import altair as alt
from vega_datasets import data

# topographical
topo_usa = 'https://vega.github.io/vega-datasets/data/us-10m.json'
topo_tx = 'https://raw.githubusercontent.com/deldersveld/topojson/master/countries/us-states/tx-48-texas-counties.json'
topo_king = 'https://raw.githubusercontent.com/johan/world.geo.json/master/countries/USA/WA/King.geo.json'

### Download Mapping County FIPS to lat, long_

In [6]:
%sh mkdir -p /dbfs/tmp/dennylee/COVID/map_fips/ && wget -O /dbfs/tmp/dennylee/COVID/map_fips/countyfips_lat_long.csv https://raw.githubusercontent.com/dennyglee/tech-talks/master/datasets/countyfips_lat_long.csv && ls -al /dbfs/tmp/dennylee/COVID/map_fips/

--2020-07-27 20:00:22-- https://raw.githubusercontent.com/dennyglee/tech-talks/master/datasets/countyfips_lat_long.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.208.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.208.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 129178 (126K) [text/plain]
Saving to: ‘/dbfs/tmp/dennylee/COVID/map_fips/countyfips_lat_long.csv’

 0K .......... .......... .......... .......... .......... 39% 2.24M 0s
 50K .......... .......... .......... .......... .......... 79% 4.90M 0s
 100K .......... .......... ...... 100% 5.12M=0.04s

2020-07-27 20:00:22 (3.35 MB/s) - ‘/dbfs/tmp/dennylee/COVID/map_fips/countyfips_lat_long.csv’ saved [129178/129178]

total 135
drwxrwxrwx 2 root root 4096 Jul 27 20:00 .
drwxrwxrwx 2 root root 4096 Jul 24 02:01 ..
-rwxrwxrwx 1 root root 129178 Jul 27 20:00 countyfips_lat_long.csv

In [7]:
# Create mapping of county FIPS to centroid long_ and lat
map_fips = spark.read.option("header", True).option("inferSchema", True).csv("/tmp/dennylee/COVID/map_fips/countyfips_lat_long.csv")
map_fips = (map_fips
              .withColumnRenamed("STATE", "state")
              .withColumnRenamed("COUNTYNAME", "county")
              .withColumnRenamed("LAT", "lat")
              .withColumnRenamed("LON", "long_"))
map_fips.createOrReplaceTempView("map_fips")

In [8]:
map_fips_dedup = spark.sql("""select fips, min(state) as state, min(county) as county, min(long_) as long_, min(lat) as lat from map_fips group by fips""")
map_fips_dedup.createOrReplaceTempView("map_fips_dedup")

### Get 2019 Population Estimates

In [10]:
%sh mkdir -p /dbfs/tmp/dennylee/COVID/population_estimates_by_county/ && wget -O /dbfs/tmp/dennylee/COVID/population_estimates_by_county/co-est2019-alldata.csv https://raw.githubusercontent.com/databricks/tech-talks/master/datasets/co-est2019-alldata.csv && ls -al /dbfs/tmp/dennylee/COVID/population_estimates_by_county/

--2020-07-27 20:00:24-- https://raw.githubusercontent.com/databricks/tech-talks/master/datasets/co-est2019-alldata.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.248.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.248.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3644730 (3.5M) [text/plain]
Saving to: ‘/dbfs/tmp/dennylee/COVID/population_estimates_by_county/co-est2019-alldata.csv’

 0K .......... .......... .......... .......... .......... 1% 3.47M 1s
 50K .......... .......... .......... .......... .......... 2% 6.04M 1s
 100K .......... .......... .......... .......... .......... 4% 5.23M 1s
 150K .......... .......... .......... .......... .......... 5% 6.04M 1s
 200K .......... .......... .......... .......... .......... 7% 6.67M 1s
 250K .......... .......... .......... .......... .......... 8% 6.77M 1s
 300K .......... .......... .......... .......... .......... 9% 7.08M 1s
 350K .......... .......... .......... .......... .......... 11% 6.82M 1s
 400K .......... .......... .......... .......... .......... 12% 6.64M 1s
 450K .......... .......... .......... .......... .......... 14% 7.19M 1s
 500K .......... .......... .......... .......... .......... 15% 7.08M 0s
 550K .......... .......... .......... .......... .......... 16% 2.71M 1s
 600K .......... .......... .......... .......... .......... 18% 6.10M 1s
 650K .......... .......... .......... .......... .......... 19% 5.42M 1s
 700K .......... .......... .......... .......... .......... 21% 14.3M 0s
 750K .......... .......... .......... .......... .......... 22% 20.8M 0s
 800K .......... .......... .......... .......... .......... 23% 15.4M 0s
 850K .......... .......... .......... .......... .......... 25% 24.9M 0s
 900K .......... .......... .......... .......... .......... 26% 23.4M 0s
 950K .......... .......... .......... .......... .......... 28% 17.9M 0s
 1000K .......... .......... .......... .......... .......... 29% 20.1M 0s
 1050K .......... .......... .......... .......... .......... 30% 22.3M 0s
 1100K .......... .......... .......... .......... .......... 32% 11.3M 0s
 1150K .......... .......... .......... .......... .......... 33% 13.9M 0s
 1200K .......... .......... .......... .......... .......... 35% 22.9M 0s
 1250K .......... .......... .......... .......... .......... 36% 19.6M 0s
 1300K .......... .......... .......... .......... .......... 37% 19.4M 0s
 1350K .......... .......... .......... .......... .......... 39% 19.8M 0s
 1400K .......... .......... .......... .......... .......... 40% 18.2M 0s
 1450K .......... .......... .......... .......... .......... 42% 18.6M 0s
 1500K .......... .......... .......... .......... .......... 43% 23.0M 0s
 1550K .......... .......... .......... .......... .......... 44% 25.2M 0s
 1600K .......... .......... .......... .......... .......... 46% 14.1M 0s
 1650K .......... .......... .......... .......... .......... 47% 22.1M 0s
 1700K .......... .......... .......... .......... .......... 49% 19.3M 0s
 1750K .......... .......... .......... .......... .......... 50% 4.58M 0s
 1800K .......... .......... .......... .......... .......... 51% 21.8M 0s
 1850K .......... .......... .......... .......... .......... 53% 19.8M 0s
 1900K .......... .......... .......... .......... .......... 54% 22.0M 0s
 1950K .......... .......... .......... .......... .......... 56% 16.7M 0s
 2000K .......... .......... .......... .......... .......... 57% 20.0M 0s
 2050K .......... .......... .......... .......... .......... 59% 18.4M 0s
 2100K .......... .......... .......... .......... .......... 60% 21.3M 0s
 2150K .......... .......... .......... .......... .......... 61% 14.4M 0s
 2200K .......... .......... .......... .......... .......... 63% 18.1M 0s
 2250K .......... .......... .......... .......... .......... 64% 11.1M 0s
 2300K .......... .......... .......... .......... .......... 66% 24.2M 0s
 2350K .......... ......

In [11]:
map_popest_county = spark.read.option("header", True).option("inferSchema", True).csv("/tmp/dennylee/COVID/population_estimates_by_county/co-est2019-alldata.csv")
map_popest_county.createOrReplaceTempView("map_popest_county")
fips_popest_county = spark.sql("select State * 1000 + substring(cast(1000 + County as string), 2, 3) as fips, STNAME, CTYNAME, census2010pop, POPESTIMATE2019 from map_popest_county")
fips_popest_county.createOrReplaceTempView("fips_popest_county")

## Specify `nyt_daily` table
* Source: `/databricks-datasets/COVID/covid-19-data/`
* Contains the COVID-19 daily reports

In [13]:
nyt_daily = spark.read.option("inferSchema", True).option("header", True).csv("/databricks-datasets/COVID/covid-19-data/us-counties.csv")
nyt_daily.createOrReplaceTempView("nyt_daily")
display(nyt_daily)

date county state fips cases deaths 2020-01-21 Snohomish Washington 53061 1 0 2020-01-22 Snohomish Washington 53061 1 0 2020-01-23 Snohomish Washington 53061 1 0 2020-01-24 Cook Illinois 17031 1 0 2020-01-24 Snohomish Washington 53061 1 0 2020-01-25 Orange California 6059 1 0 2020-01-25 Cook Illinois 17031 1 0 2020-01-25 Snohomish Washington 53061 1 0 2020-01-26 Maricopa Arizona 4013 1 0 2020-01-26 Los Angeles California 6037 1 0 2020-01-26 Orange California 6059 1 0 2020-01-26 Cook Illinois 17031 1 0 2020-01-26 Snohomish Washington 53061 1 0 2020-01-27 Maricopa Arizona 4013 1 0 2020-01-27 Los Angeles California 6037 1 0 2020-01-27 Orange California 6059 1 0 2020-01-27 Cook Illinois 17031 1 0 2020-01-27 Snohomish Washington 53061 1 0 2020-01-28 Maricopa Arizona 4013 1 0 2020-01-28 Los Angeles California 6037 1 0 2020-01-28 Orange California 6059 1 0 2020-01-28 Cook Illinois 17031 1 0 2020-01-28 Snohomish Washington 53061 1 0 2020-01-29 Maricopa Arizona 4013 1 0 2020-01-29 Los Angeles California 6037 1 0 2020-01-29 Orange California 6059 1 0 2020-01-29 Cook Illinois 17031 1 0 2020-01-29 Snohomish Washington 53061 1 0 2020-01-30 Maricopa Arizona 4013 1 0 2020-01-30 Los Angeles California 6037 1 0 2020-01-30 Orange California 6059 1 0 2020-01-30 Cook Illinois 17031 2 0 2020-01-30 Snohomish Washington 53061 1 0 2020-01-31 Maricopa Arizona 4013 1 0 2020-01-31 Los Angeles California 6037 1 0 2020-01-31 Orange California 6059 1 0 2020-01-31 Santa Clara California 6085 1 0 2020-01-31 Cook Illinois 17031 2 0 2020-01-31 Snohomish Washington 53061 1 0 2020-02-01 Maricopa Arizona 4013 1 0 2020-02-01 Los Angeles California 6037 1 0 2020-02-01 Orange California 6059 1 0 2020-02-01 Santa Clara California 6085 1 0 2020-02-01 Cook Illinois 17031 2 0 2020-02-01 Suffolk Massachusetts 25025 1 0 2020-02-01 Snohomish Washington 53061 1 0 2020-02-02 Maricopa Arizona 4013 1 0 2020-02-02 Los Angeles California 6037 1 0 2020-02-02 Orange California 6059 1 0 2020-02-02 San Francisco California 6075 2 0 2020-02-02 Santa Clara California 6085 2 0 2020-02-02 Cook Illinois 17031 2 0 2020-02-02 Suffolk Massachusetts 25025 1 0 2020-02-02 Snohomish Washington 53061 1 0 2020-02-03 Maricopa Arizona 4013 1 0 2020-02-03 Los Angeles California 6037 1 0 2020-02-03 Orange California 6059 1 0 2020-02-03 San Francisco California 6075 2 0 2020-02-03 Santa Clara California 6085 2 0 2020-02-03 Cook Illinois 17031 2 0 2020-02-03 Suffolk Massachusetts 25025 1 0 2020-02-03 Snohomish Washington 53061 1 0 2020-02-04 Maricopa Arizona 4013 1 0 2020-02-04 Los Angeles California 6037 1 0 2020-02-04 Orange California 6059 1 0 2020-02-04 San Francisco California 6075 2 0 2020-02-04 Santa Clara California 6085 2 0 2020-02-04 Cook Illinois 17031 2 0 2020-02-04 Suffolk Massachusetts 25025 1 0 2020-02-04 Snohomish Washington 53061 1 0 2020-02-05 Maricopa Arizona 4013 1 0 2020-02-05 Los Angeles California 6037 1 0 2020-02-05 Orange California 6059 1 0 2020-02-05 San Francisco California 6075 2 0 2020-02-05 Santa Clara California 6085 2 0 2020-02-05 Cook Illinois 17031 2 0 2020-02-05 Suffolk Massachusetts 25025 1 0 2020-02-05 Snohomish Washington 53061 1 0 2020-02-05 Dane Wisconsin 55025 1 0 2020-02-06 Maricopa Arizona 4013 1 0 2020-02-06 Los Angeles California 6037 1 0 2020-02-06 Orange California 6059 1 0 2020-02-06 San Francisco California 6075 2 0 2020-02-06 Santa Clara California 6085 2 0 2020-02-06 Cook Illinois 17031 2 0 2020-02-06 Suffolk Massachusetts 25025 1 0 2020-02-06 Snohomish Washington 53061 1 0 2020-02-06 Dane Wisconsin 55025 1 0 2020-02-07 Maricopa Arizona 4013 1 0 2020-02-07 Los Angeles California 6037 1 0 2020-02-07 Orange California 6059 1 0 2020-02-07 San Francisco California 6075 2 0 2020-02-07 Santa Clara California 6085 2 0 2020-02-07 Cook Illinois 17031 2 0 2020-02-07 Suffolk Massachusetts 25025 1 0 2020-02-07 Snohomish Washington 53061 1 0 2020-02-07 Dane Wisconsin 55025 1 0 2020-02-08 Maricopa Arizona 4013 1 0 2020-02-08 Los Angeles California 6037 1 0 2020-02-

# COVID-19 Cases and Deaths for Specific Counties
* Top 10 Texas counties (3/26/2020)
* Top 10 New York counties (3/18/2020)
The queries are using the US Census Population Estimates for 2019

In [15]:
# TX State 2 week window
tx_state_window = spark.sql("""
SELECT date, 100 + datediff(date, '2020-03-26T00:00:00.000+0000') as day_num, county, fips, cases, deaths, 100000.*cases/population_estimate AS cases_per_100Kpop, 100000.*deaths/population_estimate AS deaths_per_100Kpop
  from (
SELECT CAST(f.date AS date) AS date, f.county, f.fips, SUM(f.cases) AS cases, SUM(f.deaths) AS deaths, MAX(p.POPESTIMATE2019) AS population_estimate 
  FROM nyt_daily f 
    JOIN fips_popest_county p
      ON p.fips = f.fips
 WHERE f.state = 'Texas' 
   AND date BETWEEN '2020-03-26T00:00:00.000+0000' AND '2020-07-27T00:00:00.000+0000'
 GROUP BY f.date, f.county, f.fips
) a""")
tx_state_window.createOrReplaceTempView("tx_state_window")

# NY State 2 week window
fl_state_window = spark.sql("""
SELECT date, 100 + datediff(date, '2020-06-29T00:00:00.000+0000') as day_num, county, fips, cases, deaths, 100000.*cases/population_estimate AS cases_per_100Kpop, 100000.*deaths/population_estimate AS deaths_per_100Kpop
  FROM (
SELECT CAST(f.date AS date) AS date, f.county, p.fips, SUM(f.cases) as cases, SUM(f.deaths) as deaths, MAX(p.POPESTIMATE2019) AS population_estimate  
  FROM nyt_daily f 
    JOIN fips_popest_county p
      ON p.fips = coalesce(f.fips, 36061)
 WHERE f.state = 'Florida' 
   AND date BETWEEN '2020-06-22T00:00:00.000+0000' AND '2020-07-27T00:00:00.000+0000'
 GROUP BY f.date, f.county, p.fips
) a""")
fl_state_window.createOrReplaceTempView("fl_state_window")

# NY State 2 week window (-1 week)
ny_state_window_m1 = spark.sql("""
SELECT date, 100 + datediff(date, '2020-03-06T00:00:00.000+0000') as day_num, county, fips, cases, deaths, 100000.*cases/population_estimate AS cases_per_100Kpop, 100000.*deaths/population_estimate AS deaths_per_100Kpop
  FROM (
SELECT CAST(f.date AS date) AS date, f.county, p.fips, SUM(f.cases) as cases, SUM(f.deaths) as deaths, MAX(p.POPESTIMATE2019) AS population_estimate  
  FROM nyt_daily f 
    JOIN fips_popest_county p
      ON p.fips = coalesce(f.fips, 36061)
 WHERE f.state = 'New York' 
   AND date BETWEEN '2020-03-06T00:00:00.000+0000' AND '2020-03-20T00:00:00.000+0000'
 GROUP BY f.date, f.county, p.fips
) a""")
ny_state_window_m1.createOrReplaceTempView("ny_state_window_m1")

## COVID-19 Cases for TX and NY Counties

In [17]:
%sql
SELECT date, sum(cases) 
  FROM tx_state_window
  group by date
  order by date

date sum(cases) 2020-03-27 1971 2020-03-28 2381 2020-03-29 2722 2020-03-30 3065 2020-03-31 3588 2020-04-01 4402 2020-04-02 4952 2020-04-03 5773 2020-04-04 6566 2020-04-05 7117 2020-04-06 8150 2020-04-07 8968 2020-04-08 9897 2020-04-09 11372 2020-04-10 12291 2020-04-11 13208 2020-04-12 13776 2020-04-13 14487 2020-04-14 15303 2020-04-15 16106 2020-04-16 16928 2020-04-17 18131 2020-04-18 18927 2020-04-19 19441 2020-04-20 20035 2020-04-21 20949 2020-04-22 21769 2020-04-23 22706 2020-04-24 23650 2020-04-25 24494 2020-04-26 25209 2020-04-27 25963 2020-04-28 26868 2020-04-29 27839 2020-04-30 29072 2020-05-01 30342 2020-05-02 31433 2020-05-03 32299 2020-05-04 33134 2020-05-05 34286 2020-05-06 35441 2020-05-07 36682 2020-05-08 37943 2020-05-09 38849 2020-05-10 40037 2020-05-11 40988 2020-05-12 42348 2020-05-13 43822 2020-05-14 45169 2020-05-15 46987 2020-05-16 47866 2020-05-17 48828 2020-05-18 49788 2020-05-19 51080 2020-05-20 52369 2020-05-21 53441 2020-05-22 54369 2020-05-23 55451 2020-05-24 56050 2020-05-25 56490 2020-05-26 57573 2020-05-27 59105 2020-05-28 60820 2020-05-29 62073 2020-05-30 63396 2020-05-31 64974 2020-06-01 66081 2020-06-02 67966 2020-06-03 69546 2020-06-04 71330 2020-06-05 73323 2020-06-06 75077 2020-06-07 75948 2020-06-08 77326 2020-06-09 79271 2020-06-10 81771 2020-06-11 83772 2020-06-12 86031 2020-06-13 88120 2020-06-14 89542 2020-06-15 91726 2020-06-16 95973 2020-06-17 99466 2020-06-18 102677 2020-06-19 107322 2020-06-20 111594 2020-06-21 114886 2020-06-22 119646 2020-06-23 124726 2020-06-24 131310 2020-06-25 136783 2020-06-26 142582 2020-06-27 148639 2020-06-28 152866 2020-06-29 158756 2020-06-30 166703 2020-07-01 174826 2020-07-02 182260 2020-07-03 190049 2020-07-04 195707 2020-07-05 200137 2020-07-06 209350 2020-07-07 219164 2020-07-08 229116 2020-07-09 240025 2020-07-10 249948 2020-07-11 258280 2020-07-12 264811 2020-07-13 273732 2020-07-14 284638 2020-07-15 296478 2020-07-16 311516 2020-07-17 321242 2020-07-18 330059

In [18]:
%sql
SELECT f.date, f.county, f.cases 
  FROM tx_state_window f
  JOIN (
      SELECT county, sum(cases) as Cases FROM tx_state_window GROUP BY County ORDER BY cases DESC LIMIT 15
    ) x ON x.county = f.county

date county cases 2020-04-26 Galveston 541 2020-06-16 Montgomery 1359 2020-06-23 Cameron 1647 2020-07-01 Bexar 12516 2020-07-07 El Paso 7642 2020-05-16 Galveston 701 2020-06-05 Denton 1467 2020-05-21 Galveston 732 2020-05-29 Potter 2317 2020-07-12 Bexar 19660 2020-07-17 Brazoria 4359 2020-03-27 Dallas 367 2020-03-29 Montgomery 64 2020-04-22 Galveston 491 2020-06-12 Montgomery 1258 2020-06-23 Dallas 17744 2020-06-28 Harris 29276 2020-03-30 Brazoria 61 2020-04-02 Hidalgo 62 2020-06-25 Cameron 1881 2020-05-09 Galveston 680 2020-05-20 Hidalgo 467 2020-05-25 Denton 1233 2020-05-23 Cameron 686 2020-04-13 Bexar 783 2020-04-20 Bexar 1040 2020-04-25 Brazoria 445 2020-04-28 Galveston 554 2020-06-16 Fort Bend 2496 2020-06-28 Cameron 2183 2020-06-30 Denton 2845 2020-03-29 Collin 134 2020-04-05 Tarrant 404 2020-04-12 Travis 774 2020-04-18 El Paso 505 2020-04-19 Collin 527 2020-06-22 Collin 2250 2020-06-26 Tarrant 10706 2020-06-29 Tarrant 11739 2020-07-13 Hidalgo 8137 2020-07-14 Collin 4800 2020-05-06 Dallas 4869 2020-04-01 Harris 847 2020-04-14 Fort Bend 577 2020-04-24 Bexar 1206 2020-06-16 Tarrant 7642 2020-07-03 Cameron 2692 2020-05-12 Cameron 527 2020-05-30 Bexar 2837 2020-05-31 Fort Bend 1861 2020-06-05 Collin 1430 2020-07-14 Harris 49027 2020-07-14 Tarrant 19014 2020-04-02 Tarrant 325 2020-04-03 Galveston 206 2020-04-17 Tarrant 1175 2020-04-28 Cameron 376 2020-06-16 Hidalgo 1255 2020-06-24 Travis 6914 2020-06-29 El Paso 5745 2020-03-28 Hidalgo 18 2020-03-29 Brazoria 61 2020-04-26 El Paso 802 2020-06-12 Dallas 13585 2020-06-18 El Paso 4088 2020-06-23 Potter 2838 2020-05-21 Bexar 2371 2020-05-28 Denton 1318 2020-05-30 Galveston 807 2020-06-02 Brazoria 941 2020-07-09 Denton 3697 2020-07-11 Hidalgo 7669 2020-03-31 Hidalgo 38 2020-04-17 Fort Bend 695 2020-04-26 Potter 333 2020-06-09 Cameron 919 2020-04-18 Tarrant 1229 2020-04-19 Galveston 455 2020-04-28 Travis 1539 2020-04-29 Dallas 3352 2020-06-17 Brazoria 1525 2020-06-30 Montgomery 2133 2020-07-05 Dallas 25840 2020-05-10 Harris 7878 2020-05-17 Travis 2459 2020-06-04 Collin 1388 2020-05-10 Travis 2127 2020-05-22 Dallas 8477 2020-05-27 Potter 2266 2020-07-09 Harris 40919 2020-04-08 Collin 378 2020-04-16 Hidalgo 225 2020-04-23 Tarrant 1559 2020-04-26 Denton 713 2020-06-26 Potter 2858 2020-06-27 Brazoria 2064 2020-03-31 Denton 206 2020-04-08 Brazoria 184 2020-06-13 Collin 1647 2020-05-10 Potter 1124 2020-05-20 Harris 10095 2020-07-11 Collin 4459 2020-07-12 Cameron 3854 2020-04-02 Cameron 37 2020-04-07 Collin 343 2020-04-13 Brazoria 250 2020-04-30 El Paso 924 2020-06-13 El Paso 3794 2020-04-13 Harris 3747 2020-04-15 Denton 521 2020-05-01 Travis 1683 2020-06-21 Harris 21053 2020-06-26 Hidalgo 2871 2020-07-01 Harris 32859 2020-05-12 El Paso 1413 2020-07-15 Bexar 21558 2020-07-18 Montgomery 4721 2020-04-19 El Paso 531 2020-06-30 Dallas 21338 2020-07-02 El Paso 6389 2020-07-05 Collin 3521 2020-05-11 Galveston 686 2020-05-14 Fort Bend 1471 2020-05-16 Brazoria 727 2020-05-19 Bexar 2278 2020-06-03 Hidalgo 610 2020-07-18 Potter 3159 2020-04-19 Dallas 2428 2020-06-14 Montgomery 1258 2020-06-19 Galveston 1642 2020-05-09 Cameron 494 2020-05-17 Hidalgo 440 2020-04-01 Potter 16 2020-06-19 Travis 5286 2020-04-18 Cameron 298 2020-04-30 Bexar 1374 2020-06-07 Galveston 994 2020-06-23 Montgomery 1737 2020-07-08 Cameron 3120 2020-05-29 Collin 1278 2020-06-15 Potter 2789 2020-07-06 Fort Bend 4286 2020-07-06 Galveston 4586 2020-07-12 Hidalgo 7982 2020-05-14 Cameron 582 2020-05-26 Travis 2984 2020-05-27 Harris 11542 2020-06-03 Bexar 2965 2020-07-15 Harris 50370 2020-03-27 Fort Bend 101 2020-04-07 Cameron 101 2020-04-09 Dallas 1432 2020-04-09 Hidalgo 139 2020-06-24 Montgomery 1831 2020-07-08 Fort Bend 4413 2020-04-03 Fort Bend 221 2020-04-06 Montgomery 161 2020-04-08 Galveston 287 2020-04-24 Galveston 522 2020-04-25 Potter 326 2020-04-27 Fort Bend 957 2020-04-29 Montgomery 582 2020-06-18 Collin 2026 2020-06-22 Hidalgo 1882 2020-07-14 Cameron 4175 2020-05-23 Potter 2238 2020-06-05 Dallas 11541 2020-05-09 Harris 7803

In [19]:
%sql
SELECT f.date, f.county, f.cases 
  FROM ny_state_window f
  JOIN (
      SELECT county, sum(cases) as cases FROM ny_state_window GROUP BY county ORDER BY cases DESC LIMIT 10
    ) x ON x.county = f.county

com.databricks.backend.common.rpc.DatabricksExceptions$SQLExecutionException: org.apache.spark.sql.AnalysisException: Table or view not found: ny_state_window; line 2 pos 7;
'Project ['f.date, 'f.county, 'f.cases]
+- 'Join Inner, ('x.county = 'f.county)
 :- 'SubqueryAlias f
 : +- 'UnresolvedRelation [ny_state_window]
 +- 'SubqueryAlias x
 +- 'GlobalLimit 10
 +- 'LocalLimit 10
 +- 'Sort ['cases DESC NULLS LAST], true
 +- 'Aggregate ['county], ['county, 'sum('cases) AS cases#14433]
 +- 'UnresolvedRelation [ny_state_window]

	at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:42)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1(CheckAnalysis.scala:109)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1$adapted(CheckAnalysis.scala:95)
	at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:184)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1$adapted(TreeNode.scala:183)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1$adapted(TreeNode.scala:183)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$foreachUp$1$adapted(TreeNode.scala:183)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:183)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.checkAnalysis(CheckAnalysis.scala:95)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.checkAnalysis$(CheckAnalysis.scala:92)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.checkAnalysis(Analyzer.scala:136)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$executeAndCheck$1(Analyzer.scala:164)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.markInAnalyzer(AnalysisHelper.scala:201)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.executeAndCheck(Analyzer.scala:161)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$analyzed$1(QueryExecution.scala:90)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:162)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:828)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:162)
	at org.apache.spark.sql.execution.QueryExecution.analyzed$lzycompute(QueryExecution.scala:90)
	at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:88)
	at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:80)
	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:102)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:828)
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:100)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:663)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:828)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:658)
	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:672)
	at com.databricks.backend.daemon.driver.SQLDriverLocal.$anonfun$executeSql$1(SQLDriverLocal.scala:88)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at sc

In [20]:
%sql
SELECT f.date, f.county, f.cases 
  FROM ny_state_window_m1 f
  JOIN (
      SELECT county, sum(cases) as cases FROM ny_state_window_m1 GROUP BY county ORDER BY cases DESC LIMIT 10
    ) x ON x.county = f.county

## COVID-19 Cases per 100K people for TX and NY Counties
The concern with the above graphs is while it provides you the total number of cases, it's hard to compare TX state and NY state due to density.  While not perfect, a better way to look at these numbers would be to review this data as a proportion of population estimates.

Let's look at these values by a percentage of the population; the numbers used are the 2019 US Census estimates of county populations.

*Note, reviewing the top 10 counties by case (vs. % of cases)*

In [22]:
%sql
SELECT f.date, f.county, f.cases_per_100Kpop
  FROM tx_state_window f
  JOIN (
      SELECT county, sum(cases) as cases FROM tx_state_window GROUP BY county ORDER BY cases DESC LIMIT 15  
    ) x ON x.county = f.county

date county cases_per_100Kpop 2020-04-26 Galveston 158.122868191 2020-06-16 Montgomery 223.74384869055 2020-06-23 Cameron 389.21172219688 2020-07-01 Bexar 624.6899260015 2020-07-07 El Paso 910.58793810576 2020-05-16 Galveston 204.88748724933 2020-06-05 Denton 165.35036355664 2020-03-27 Dallas 13.9251668364 2020-03-29 Montgomery 10.53686998984 2020-04-22 Galveston 143.50892473527 2020-06-12 Montgomery 207.11535073783 2020-06-23 Dallas 673.26474208466 2020-06-28 Harris 621.13263990919 2020-07-12 Bexar 981.2563075415 2020-07-17 Brazoria 1164.68589017378 2020-05-21 Galveston 213.94813219189 2020-05-29 Potter 1973.34241791935 2020-03-30 Brazoria 16.29865549452 2020-04-02 Hidalgo 7.13704390548 2020-06-25 Cameron 444.50956250901 2020-05-09 Galveston 198.74963099793 2020-05-20 Hidalgo 53.75805651388 2020-05-25 Denton 138.97545894025 2020-04-13 Bexar 39.0805538558 2020-04-20 Bexar 51.90775991064 2020-04-25 Brazoria 118.90002778787 2020-04-28 Galveston 161.92249348949 2020-06-16 Fort Bend 307.50731808281 2020-06-28 Cameron 515.87686068962 2020-06-30 Denton 320.66924629765 2020-05-23 Cameron 162.11247202615 2020-03-29 Collin 12.9502382264 2020-04-05 Tarrant 19.21508288883 2020-04-12 Travis 60.75572587393 2020-04-18 El Paso 60.17363370105 2020-04-19 Collin 50.93116078591 2020-06-22 Collin 217.44802992085 2020-06-26 Tarrant 509.19969655389 2020-06-29 Tarrant 558.3313317622 2020-07-13 Hidalgo 936.67945578889 2020-07-14 Collin 463.88913049781 2020-05-06 Dallas 184.74560579408 2020-05-12 Cameron 124.53829848073 2020-05-30 Bexar 141.59837967931 2020-05-31 Fort Bend 229.27528804171 2020-04-01 Harris 17.97032880185 2020-04-14 Fort Bend 71.08642729719 2020-04-24 Bexar 60.1930369733 2020-06-16 Tarrant 363.46946395151 2020-07-03 Cameron 636.16147914633 2020-06-05 Collin 138.20030346081 2020-04-02 Tarrant 15.45767806651 2020-04-03 Galveston 60.20944703761 2020-04-17 Tarrant 55.88545147121 2020-04-28 Cameron 88.85464939042 2020-06-16 Hidalgo 144.46758228033 2020-06-24 Travis 542.71975283252 2020-06-29 El Paso 684.54955566836 2020-07-14 Harris 1040.17864246578 2020-07-14 Tarrant 904.34551002014 2020-03-28 Hidalgo 2.07204500482 2020-03-29 Brazoria 16.29865549452 2020-04-26 El Paso 95.56287965988 2020-06-12 Dallas 515.4588323501 2020-06-18 El Paso 487.10854370274 2020-06-23 Potter 2417.06766597113 2020-03-31 Hidalgo 4.37431723239 2020-04-17 Fort Bend 85.62403287963 2020-04-26 Potter 283.60941958012 2020-06-09 Cameron 217.17399678138 2020-07-09 Denton 416.70095028556 2020-07-11 Hidalgo 882.80628566364 2020-05-21 Bexar 118.33971033474 2020-05-28 Denton 148.55608668552 2020-05-30 Galveston 235.86904737548 2020-06-02 Brazoria 251.42680033345 2020-04-18 Tarrant 58.45380413457 2020-04-19 Galveston 132.98688544714 2020-04-28 Travis 120.8049898191 2020-04-29 Dallas 127.18571998804 2020-06-17 Brazoria 407.46638736293 2020-06-30 Montgomery 351.1741201302 2020-07-05 Dallas 980.45316363096 2020-05-10 Harris 167.14315265762 2020-05-17 Travis 193.02109809302 2020-06-04 Collin 134.14127356895 2020-04-08 Collin 36.5312690267 2020-04-16 Hidalgo 25.90056256022 2020-04-23 Tarrant 74.14929263287 2020-04-26 Denton 80.3645597927 2020-06-26 Potter 2434.10126474471 2020-06-27 Brazoria 551.48237607678 2020-07-09 Harris 868.15570748887 2020-05-10 Travis 166.96050249852 2020-05-22 Dallas 321.64479365711 2020-05-27 Potter 1929.90674104671 2020-03-31 Denton 23.21893312384 2020-04-08 Brazoria 49.16315755723 2020-06-13 Collin 159.17195790206 2020-05-10 Potter 957.28825107525 2020-05-20 Harris 214.18001092647 2020-04-02 Cameron 8.74367560491 2020-04-07 Collin 33.14874411682 2020-04-13 Brazoria 66.79776842015 2020-04-30 El Paso 110.09987631637 2020-06-13 El Paso 452.0767648748 2020-07-11 Collin 430.9336735187 2020-07-12 Cameron 910.76015625185 2020-04-13 Harris 79.49801891446 2020-04-15 Denton 58.72361241514 2020-05-01 Travis 132.10838067937 2020-06-21 Harris 446.66981377266 2020-06-26 Hidalgo 330.49117826839 2020-07-01 Harris 697.15116186556 2020-05-12 El Paso 168.36701865263 2020-04-

In [23]:
%sql
SELECT f.date, f.county, f.cases_per_100Kpop
  FROM tx_state_window f
  JOIN (
      SELECT county, sum(cases) as cases FROM tx_state_window GROUP BY county ORDER BY cases DESC LIMIT 15  
    ) x ON x.county = f.county
      WHERE x.county = "Bexar"

In [24]:
%sql
SELECT f.date, f.county, f.cases_per_100Kpop 
  FROM ny_state_window f
  JOIN (
      SELECT county, sum(cases) as cases FROM ny_state_window GROUP BY county ORDER BY cases DESC LIMIT 10
    ) x ON x.county = f.county

In [25]:
%sql
SELECT f.date, f.county, f.cases_per_100Kpop 
  FROM ny_state_window_m1 f
  JOIN (
      SELECT county, sum(cases) as cases FROM ny_state_window_m1 GROUP BY county ORDER BY cases DESC LIMIT 10
    ) x ON x.county = f.county

## Visualize Cases by State Choropleth Maps
* Join the data with `map_fips_dedup` to obtain the county centroid lat, long_

In [27]:
# Extract Day Number and county centroid lat, long_
tx_daynum = spark.sql("""select f.fips, f.county, f.date, f.day_num, cases as confirmed, cast(f.cases_per_100Kpop as int) as confirmed_per100K, deaths, cast(f.deaths_per_100Kpop as int) as deaths_per100K, m.lat, m.long_ from tx_state_window f join map_fips_dedup m on m.fips = f.fips""")
tx_daynum.createOrReplaceTempView("tx_daynum")
fl_daynum = spark.sql("""select cast(f.fips as int) as fips, f.county, f.date, f.day_num, cases as confirmed, cast(f.cases_per_100Kpop as int) as confirmed_per100K, deaths, cast(f.deaths_per_100Kpop as int) as deaths_per100K, m.lat, m.long_ from fl_state_window f join map_fips_dedup m on m.fips = f.fips""")
fl_daynum.createOrReplaceTempView("fl_daynum")
ny_daynum_m1 = spark.sql("""select cast(f.fips as int) as fips, f.county, f.date, f.day_num, cases as confirmed, cast(f.cases_per_100Kpop as int) as confirmed_per100K, deaths, cast(f.deaths_per_100Kpop as int) as deaths_per100K, m.lat, m.long_ from ny_state_window_m1 f join map_fips_dedup m on m.fips = f.fips""")
ny_daynum_m1.createOrReplaceTempView("ny_daynum_m1")

In [28]:
display(tx_daynum)

fips county date day_num confirmed confirmed_per100K deaths deaths_per100K lat long_ 48275 Knox 2020-05-02 137 1 27 0 0 33.6061 -99.7414 48503 Young 2020-05-03 138 4 22 1 5 33.1767 -98.6878 48237 Jack 2020-05-04 139 4 44 0 0 33.2335 -98.1725 48037 Bowie 2020-05-05 140 104 111 10 10 33.4454 -94.4229 48171 Gillespie 2020-05-05 140 4 14 0 0 30.3181 -98.9464 48217 Hill 2020-05-05 140 18 49 1 2 31.9907 -97.1324 48069 Castro 2020-05-06 141 20 265 1 13 34.53 -102.2617 48277 Lamar 2020-05-06 141 71 142 0 0 33.6671 -95.5711 48175 Goliad 2020-05-07 142 7 91 0 0 28.6571 -97.4264 48119 Delta 2020-05-08 143 1 18 0 0 33.3863 -95.6723 48177 Gonzales 2020-05-09 144 40 191 2 9 29.4567 -97.4925 48199 Hardin 2020-05-10 145 115 199 4 6 30.3324 -94.3902 48405 San Augustine 2020-05-10 145 21 254 1 12 31.3943 -94.1682 48267 Kimble 2020-05-11 146 1 23 0 0 30.4868 -99.7487 48147 Fannin 2020-05-12 147 23 64 1 2 33.5939 -96.1066 48041 Brazos 2020-05-14 149 305 133 18 7 30.6611 -96.3023 48359 Oldham 2020-05-14 149 3 142 1 47 35.405 -102.6029 48165 Gaines 2020-05-15 150 5 23 0 0 32.7407 -102.6352 48343 Morris 2020-05-15 150 10 80 0 0 33.1135 -94.7326 48167 Galveston 2020-05-16 151 701 204 31 9 29.3941 -94.965 48221 Hood 2020-05-16 151 22 35 3 4 32.4299 -97.8323 48477 Washington 2020-05-16 151 181 504 22 61 30.2145 -96.4036 48491 Williamson 2020-05-16 151 463 78 19 3 30.6481 -97.6008 48115 Dawson 2020-05-17 152 38 298 1 7 32.7425 -101.9477 48289 Leon 2020-05-18 153 9 51 0 0 31.2965 -95.9957 48187 Guadalupe 2020-05-19 154 110 65 0 0 29.5831 -97.9484 48463 Uvalde 2020-05-19 154 6 22 0 0 29.3574 -99.7622 48177 Gonzales 2020-05-20 155 88 422 2 9 29.4567 -97.4925 48359 Oldham 2020-05-20 155 3 142 1 47 35.405 -102.6029 48469 Victoria 2020-05-21 156 157 170 7 7 28.7964 -96.9715 48249 Jim Wells 2020-05-22 157 9 22 0 0 27.7311 -98.0902 48287 Lee 2020-05-22 157 8 46 0 0 30.3107 -96.9657 48473 Waller 2020-05-25 160 53 95 0 0 30.011 -95.9877 48467 Van Zandt 2020-05-26 161 26 45 1 1 32.5637 -95.8365 48487 Wilbarger 2020-05-26 161 6 46 0 0 34.0807 -99.241 48247 Jim Hogg 2020-05-27 162 4 76 0 0 27.0434 -98.6973 48277 Lamar 2020-05-29 164 138 276 9 18 33.6671 -95.5711 48149 Fayette 2020-05-30 165 34 134 2 7 29.8768 -96.9198 48199 Hardin 2020-05-31 166 138 239 5 8 30.3324 -94.3902 48307 McCulloch 2020-05-31 166 3 37 0 0 31.1989 -99.3474 48093 Comanche 2020-06-01 167 12 88 1 7 31.948 -98.5583 48265 Kerr 2020-06-01 167 19 36 0 0 30.0615 -99.3501 48321 Matagorda 2020-06-02 168 70 191 5 13 28.8661 -95.9853 48307 McCulloch 2020-06-02 168 3 37 0 0 31.1989 -99.3474 48481 Wharton 2020-06-02 168 64 154 0 0 29.2779 -96.2221 48129 Donley 2020-06-03 169 26 793 0 0 34.9654 -100.814 48265 Kerr 2020-06-03 169 20 38 0 0 30.0615 -99.3501 48371 Pecos 2020-06-03 169 103 650 0 0 30.7811 -102.7236 48503 Young 2020-06-03 169 4 22 1 5 33.1767 -98.6878 48169 Garza 2020-06-04 170 6 96 0 0 33.1799 -101.2984 48117 Deaf Smith 2020-06-05 171 174 938 13 70 34.966 -102.6049 48121 Denton 2020-06-05 171 1467 165 33 3 33.2043 -97.1171 48251 Johnson 2020-05-29 164 195 110 4 2 32.379 -97.3663 48471 Walker 2020-07-09 205 2172 2976 28 38 30.739 -95.5722 48297 Live Oak 2020-07-10 206 65 532 0 0 28.3514 -98.1248 48457 Tyler 2020-07-10 206 36 166 0 0 30.7713 -94.3765 48467 Van Zandt 2020-07-12 208 182 321 3 5 32.5637 -95.8365 48241 Jasper 2020-07-13 209 71 199 2 5 30.744 -94.025 48311 McMullen 2020-07-17 213 1 134 0 0 28.3491 -98.5674 48423 Smith 2020-07-17 213 1763 757 11 4 32.375 -95.2692 48165 Gaines 2020-07-18 214 57 265 0 0 32.7407 -102.6352 48183 Gregg 2020-07-18 214 1034 834 16 12 32.4803 -94.8172 48351 Newton 2020-07-18 214 28 205 0 0 30.7863 -93.7448 48123 DeWitt 2020-03-31 105 3 14 0 0 29.0821 -97.3568 48189 Hale 2020-03-31 105 1 2 0 0 34.0705 -101.8269 48279 Lamb 2020-04-03 108 1 7 0 0 34.0686 -102.3517 48273 Kleberg 2020-04-06 111 2 6 0 0 27.4338 -97.7344 48147 Fannin 2020-04-07 112 2 5 0 0 33.5939 -96.1066 48115 Dawson 2020-04-09 114 5 39 1 7 32.7425 -101.9477 48195 Hansford 2020-04-10 115 1 18 0 0 3

In [29]:
from pyspark.sql.functions import weekofyear, month
 
dfmax = tx_daynum.withColumn('week_of_year',weekofyear(tx_daynum.date)).withColumn('month',month(tx_daynum.date))
dfmax = dfmax.where(col('county').isin({'Bexar', 'Brazoria', 'Cameron', 'Collin', 'Dallas', 'Denton', 'El Paso', 'Fort Bend', 'Galveston', 'Harris', 'Hidalgo', 'Montgomery', 'Potter', 'Tarrant', 'Travis'}))

In [30]:
from pyspark.sql.functions import avg, stddev, col
dfmax = dfmax.select("county","week_of_year","deaths_per100K").groupBy("county","week_of_year").agg(avg(col("deaths_per100K")).alias('mean'),stddev(col("deaths_per100K")).alias('stdev'))
cvdf = dfmax.withColumn('cov',col("stdev")/col("mean")).na.drop()

In [31]:
from pyspark.sql.functions import asc, desc
display(cvdf.filter(cvdf.week_of_year >18).sort(asc('week_of_year')))

county week_of_year mean stdev cov Dallas 19 4.428571428571429 0.5345224838248487 0.12069862537980454 Collin 19 2.0 0.0 0.0 Brazoria 19 1.5714285714285714 0.5345224838248488 0.3401506715249038 Travis 19 4.0 0.0 0.0 Bexar 19 2.0 0.0 0.0 Denton 19 2.0 0.0 0.0 Fort Bend 19 3.5714285714285716 0.5345224838248487 0.1496662954709576 Tarrant 19 3.7142857142857144 0.48795003647426655 0.13137116366614868 Cameron 19 4.285714285714286 0.4879500364742665 0.11385500851066219 El Paso 19 2.5714285714285716 0.5345224838248487 0.20786985482077447 Potter 19 11.0 1.914854215512676 0.17407765595569782 Galveston 19 8.0 0.0 0.0 Montgomery 19 2.0 0.0 0.0 Harris 19 2.857142857142857 0.3779644730092273 0.13228756555322954 Bexar 20 2.4285714285714284 0.5345224838248488 0.22009749333964365 Galveston 20 8.285714285714286 0.48795003647426677 0.058890521643445985 Fort Bend 20 4.0 0.0 0.0 Potter 20 17.285714285714285 2.3603873774083297 0.1365513358831265 Travis 20 5.428571428571429 0.5345224838248487 0.09846466807299843 Cameron 20 5.571428571428571 0.5345224838248486 0.0959399329942036 Brazoria 20 2.0 0.0 0.0 Dallas 20 5.571428571428571 0.5345224838248487 0.09593993299420361 Montgomery 20 2.0 0.0 0.0 Denton 20 2.5714285714285716 0.5345224838248487 0.20786985482077447 Collin 20 2.0 0.0 0.0 Tarrant 20 4.857142857142857 0.37796447300922725 0.0778162150313115 El Paso 20 4.285714285714286 0.9511897312113419 0.22194427061597977 Hidalgo 20 0.7142857142857143 0.4879500364742666 0.6831300510639732 Harris 20 3.5714285714285716 0.5345224838248488 0.14966629547095767 Travis 21 6.0 0.0 0.0 Montgomery 21 2.7142857142857144 0.48795003647426655 0.1797710660694666 El Paso 21 5.857142857142857 0.37796447300922725 0.0645305197820632 Fort Bend 21 4.571428571428571 0.5345224838248487 0.11692679333668565 Denton 21 3.0 0.0 0.0 Galveston 21 9.0 0.0 0.0 Collin 21 2.4285714285714284 0.5345224838248488 0.22009749333964365 Cameron 21 6.857142857142857 0.3779644730092272 0.055119818980512304 Tarrant 21 5.857142857142857 0.37796447300922725 0.0645305197820632 Bexar 21 3.0 0.0 0.0 Hidalgo 21 1.0 0.0 0.0 Harris 21 4.0 0.0 0.0 Dallas 21 7.0 0.5773502691896257 0.08247860988423225 Potter 21 20.857142857142858 1.2149857925879117 0.05825274348024234 Brazoria 21 2.4285714285714284 0.5345224838248487 0.2200974933396436 El Paso 22 8.0 1.1547005383792517 0.14433756729740646 Collin 22 3.0 0.0 0.0 Montgomery 22 3.7142857142857144 0.48795003647426655 0.13137116366614868 Tarrant 22 6.714285714285714 0.4879500364742667 0.07267340968765675 Fort Bend 22 5.0 0.0 0.0 Potter 22 23.142857142857142 1.3451854182690983 0.05812529585113388 Brazoria 22 3.0 0.0 0.0 Dallas 22 8.0 0.0 0.0 Galveston 22 9.285714285714286 0.4879500364742667 0.052548465466459485 Bexar 22 3.0 0.0 0.0 Cameron 22 7.428571428571429 0.5345224838248487 0.07195494974565271 Travis 22 6.714285714285714 0.4879500364742667 0.07267340968765675 Harris 22 4.0 0.0 0.0 Hidalgo 22 1.0 0.0 0.0 Denton 22 3.0 0.0 0.0 Hidalgo 23 1.0 0.0 0.0 El Paso 23 9.714285714285714 0.4879500364742667 0.05023015081352746 Tarrant 23 7.571428571428571 0.5345224838248486 0.07059730918441397 Fort Bend 23 5.0 0.0 0.0 Collin 23 3.0 0.0 0.0 Brazoria 23 3.0 0.0 0.0 Denton 23 3.0 0.0 0.0 Potter 23 26.285714285714285 0.7559289460184546 0.028758166424615123 Bexar 23 3.0 0.0 0.0 Montgomery 23 4.0 0.0 0.0 Dallas 23 9.0 0.5773502691896257 0.06415002990995841 Travis 23 7.0 0.0 0.0 Harris 23 4.857142857142857 0.37796447300922725 0.0778162150313115 Cameron 23 8.571428571428571 0.5345224838248487 0.062360956446232345 Galveston 23 10.571428571428571 0.5345224838248492 0.05056293765910736 Collin 24 3.0 0.0 0.0 Galveston 24 11.0 0.0 0.0 Travis 24 7.428571428571429 0.5345224838248486 0.0719549497456527 Tarrant 24 8.571428571428571 0.5345224838248486 0.06236095644623234 Dallas 24 10.0 0.0 0.0 Harris 24 5.0 0.0 0.0 Fort Bend 24 5.285714285714286 0.48795003647426655 0.09231487176540178 El Paso 24 10.857142857142858 0.3779644730092272 0.03481251725084987 Cameron 24 9.428571428571429 0.534522483824

In [32]:
from pyspark.sql.functions import avg, stddev, col
dfmax1 = dfmax.select("county","month","confirmed_per100K").groupBy("county","month").agg(avg(col("confirmed_per100K")).alias('mean'),stddev(col("confirmed_per100K")).alias('stdev'))
cvdf1 = dfmax1.withColumn('cov',col("stdev")/col("mean")).na.drop()

In [33]:
cvdf2 = cvdf1.replace("3","March")
cvdf2 = cvdf1.replace("4","April")
cvdf2 = cvdf1.replace("5","May")
cvdf2 = cvdf1.replace("6","June")
cvdf2 = cvdf1.replace("7","July")

In [34]:
from pyspark.sql.functions import *
cvdf2 = cvdf1.withColumn('MonthName', regexp_replace('month', '3', '3 March'))
cvdf3 = cvdf2.withColumn('MonthName', regexp_replace('MonthName', '4', '4 April'))
cvdf4 = cvdf3.withColumn('MonthName', regexp_replace('MonthName', '5', '5 May'))
cvdf5 = cvdf4.withColumn('MonthName', regexp_replace('MonthName', '6', '6 June'))
cvdf6 = cvdf5.withColumn('MonthName', regexp_replace('MonthName', '7', '7 July'))

In [35]:
display(cvdf6.sort(asc('MonthName')))

county month mean stdev cov MonthName El Paso 3 4.2 0.8366600265340756 0.19920476822239894 3 March Brazoria 3 16.0 5.612486080160912 0.350780380010057 3 March Montgomery 3 9.4 2.5099800796022267 0.26701915740449217 3 March Cameron 3 4.0 1.224744871391589 0.30618621784789724 3 March Collin 3 13.0 1.8708286933869709 0.14390989949130545 3 March Hidalgo 3 2.8 0.8366600265340756 0.29880715233359845 3 March Dallas 3 17.6 3.9115214431215892 0.22224553654099938 3 March Harris 3 9.8 3.701351104664349 0.377688888231056 3 March Travis 3 14.0 2.0 0.14285714285714285 3 March Bexar 3 6.8 1.9235384061671343 0.2828732950245786 3 March Potter 3 2.6 1.51657508881031 0.5832981110808885 3 March Fort Bend 3 13.8 2.04939015319192 0.14850653283999418 3 March Galveston 3 21.4 6.54217089351845 0.3057089202578715 3 March Denton 3 18.6 3.361547262794322 0.18072834746206032 3 March Tarrant 3 6.0 1.224744871391589 0.20412414523193148 3 March Travis 4 73.76666666666667 33.07951062196213 0.4484343961404717 4 April Cameron 4 53.333333333333336 28.70099521683493 0.5381436603156549 4 April Brazoria 4 76.4 33.53186852676945 0.4388988027064064 4 April Bexar 4 42.3 17.51285242329187 0.414015423718484 4 April Dallas 4 77.76666666666667 31.32112136686937 0.402757668669559 4 April Harris 4 82.23333333333333 36.91339662744019 0.44888605546137245 4 April Tarrant 4 52.13333333333333 29.187542611157948 0.5598633493188865 4 April Montgomery 4 54.06666666666667 25.52339472250742 0.47207265208090166 4 April Galveston 4 114.86666666666666 39.12996323776073 0.34065551280697093 4 April Potter 4 132.83333333333334 130.16040060270814 0.9798775453152431 4 April Denton 4 58.5 18.37867955183303 0.3141654624244963 4 April Fort Bend 4 77.26666666666667 33.95527822799027 0.4394557147712287 4 April Collin 4 45.96666666666667 15.239430103683583 0.3315321995000054 4 April El Paso 4 51.0 33.49009117480201 0.6566684544078826 4 April Hidalgo 4 23.433333333333334 10.287868092323963 0.43902708786588746 4 April Montgomery 5 130.5483870967742 18.158264802705162 0.13909221865180627 5 May Brazoria 5 195.93548387096774 31.947806897992194 0.1630526858475071 5 May Galveston 5 207.1290322580645 15.461224478209719 0.07464537592656927 5 May Tarrant 5 195.61290322580646 47.961566154130004 0.24518610665864612 5 May Travis 5 192.09677419354838 37.93007148135428 0.19745293298438 5 May Cameron 5 139.0 26.338817994233022 0.1894878992390865 5 May Harris 5 196.5483870967742 38.54853106555201 0.19612743525884002 5 May Hidalgo 5 50.03225806451613 8.344594541648869 0.16678428806648288 5 May Dallas 5 269.1290322580645 74.98032716896876 0.278603636849818 5 May El Paso 5 209.19354838709677 68.93930632802486 0.3295479562326554 5 May Bexar 5 106.2258064516129 18.593385342498113 0.1750364244207232 5 May Potter 5 1442.8387096774193 522.1707956070948 0.36190517491995927 5 May Denton 5 119.03225806451613 20.922848548843664 0.17577460840491968 5 May Collin 5 98.51612903225806 15.62022400125755 0.1585549914993399 5 May Fort Bend 5 185.67741935483872 27.36468173488618 0.14737754235258366 5 May Harris 6 416.06666666666666 125.79236678046043 0.3023370456187961 6 June Bexar 6 282.93333333333334 133.49155554677924 0.4718127552313121 6 June Galveston 6 474.4 217.09073266009435 0.45761115653476886 6 June El Paso 6 487.1666666666667 105.23834470098008 0.21602123441870696 6 June Denton 6 218.5 51.805837775160114 0.23709765572155658 6 June Brazoria 6 382.6666666666667 120.99653267569377 0.3161930296403147 6 June Fort Bend 6 332.03333333333336 76.65708835719174 0.23087166456337235 6 June Cameron 6 312.53333333333336 115.52391402927654 0.36963709693667834 6 June Dallas 6 573.6666666666666 122.64620643959185 0.21379350338104333 6 June Collin 6 187.86666666666667 49.28644867495025 0.2623480234649587 6 June Hidalgo 6 177.3 112.33728462643816 0.6336000260938418 6 June Tarrant 6 382.9 96.14405499220857 0.2510944241112786 6 June Potter 6 2351.7 114.95190898506364 0.04888034570100933 6 June Travis 6 417.43333333333334 139.66281645037301 0.33

In [36]:
# Obtain Topography
topo_usa = 'https://vega.github.io/vega-datasets/data/us-10m.json'
topo_tx = 'https://raw.githubusercontent.com/deldersveld/topojson/master/countries/us-states/tx-48-texas-counties.json'
topo_ny = 'https://raw.githubusercontent.com/deldersveld/topojson/master/countries/us-states/NY-36-new-york-counties.json'
topo_fl = 'https://raw.githubusercontent.com/deldersveld/topojson/master/countries/us-states/FL-12-florida-counties.json'
us_counties = alt.topo_feature(topo_usa, 'counties')
tx_counties = alt.topo_feature(topo_tx, 'cb_2015_texas_county_20m')
ny_counties = alt.topo_feature(topo_ny, 'cb_2015_new_york_county_20m')
fl_counties = alt.topo_feature(topo_fl, 'cb_2015_florida_county_20m')

In [37]:
# Review TX
confirmed_tx = tx_daynum.select("fips", "day_num", "date", "confirmed", "confirmed_per100K", "county").where("confirmed > 0").toPandas()
confirmed_tx['date'] = confirmed_tx['date'].astype(str)
deaths_tx = tx_daynum.select("lat", "long_", "day_num", "date", "deaths", "deaths_per100K", "county").where("deaths > 0").toPandas()
deaths_tx['date'] = deaths_tx['date'].astype(str)

# Review FL
confirmed_fl = fl_daynum.select("fips", "day_num", "date", "confirmed", "confirmed_per100K", "county").where("confirmed > 0").toPandas()
confirmed_fl['date'] = confirmed_fl['date'].astype(str)
deaths_fl = fl_daynum.select("lat", "long_", "day_num", "date", "deaths", "deaths_per100K", "county").where("deaths > 0").toPandas()
deaths_fl['date'] = deaths_fl['date'].astype(str)

# Review NY -1 week
confirmed_ny_m1 = ny_daynum_m1.select("fips", "day_num", "date", "confirmed", "confirmed_per100K", "county").where("confirmed > 0").toPandas()
confirmed_ny_m1['date'] = confirmed_ny_m1['date'].astype(str)
deaths_ny_m1 = ny_daynum_m1.select("lat", "long_", "day_num", "date", "deaths", "deaths_per100K", "county").where("deaths > 0").toPandas()
deaths_ny_m1['date'] = deaths_ny_m1['date'].astype(str)

In [38]:
# State Choropleth Map Visualization Function
def map_state(curr_day_num, state_txt, state_counties, confirmed, confirmed_min, confirmed_max, deaths, deaths_min, deaths_max):
  # Get date_str
  date_str = confirmed[confirmed['day_num'] == 101]['date'].head(1).item()
  
  # State
  base_state = alt.Chart(state_counties).mark_geoshape(
      fill='white',
      stroke='lightgray',
  ).properties(
      width=800,
      height=600,
  ).project(
      type='mercator'
  )

  # counties
  base_state_counties = alt.Chart(us_counties).mark_geoshape(
  ).transform_lookup(
    lookup='id',
    from_=alt.LookupData(confirmed[(confirmed['confirmed_per100K'] > 0) & (confirmed['day_num'] == curr_day_num)], 'fips', ['confirmed_per100K', 'confirmed', 'county', 'date', 'fips'])  
  ).encode(
     color=alt.Color('confirmed_per100K:Q', scale=alt.Scale(type='log', domain=[confirmed_min, confirmed_max]), title='Confirmed per 100K'),
    tooltip=[
      alt.Tooltip('fips:O'),
      alt.Tooltip('confirmed:Q'),
      alt.Tooltip('confirmed_per100K:Q'),
      alt.Tooltip('county:N'),
      alt.Tooltip('date:N'),
    ],
  )

  # deaths by long, latitude
  points = alt.Chart(deaths[(deaths['deaths_per100K'] > 0) & (deaths['day_num'] == curr_day_num)]).mark_point(opacity=0.75, filled=True).encode(
    longitude='long_:Q',
    latitude='lat:Q',
    size=alt.Size('sum(deaths_per100K):Q', scale=alt.Scale(type='symlog', domain=[deaths_min, deaths_max]), title='Deaths per 100K'),
    color=alt.value('#BD595D'),
    stroke=alt.value('brown'),
    tooltip=[
      alt.Tooltip('lat'),
      alt.Tooltip('long_'),
      alt.Tooltip('deaths'),
      alt.Tooltip('county:N'),      
      alt.Tooltip('date:N'),      
    ],
  ).properties(
    # update figure title
    title=f'COVID-19 {state_txt} Confirmed Cases and Deaths per 100K by County'# [{curr_day_num}, {date_str}]'
  )

  return (base_state + base_state_counties + points)

### Texas (cases and deaths per 100K)

| Variables| Date | Cases | Deaths |
| ------- | -- | 
| Day 00 | 3/26/2020 | 1,396 | 18 |
| Day 21 | 4/16/2020 | 16,455 | 393 |
| Day 42 | 5/07/2020 | 35,390 | 973 |
| Day 63 | 5/28/2020 | 59,776 | 1,601 |
| Day 84 | 6/18/2020 | 99,851 | 2,105 |
| Day 105 | 7/09/2020 | 230,346 | 2,918 |
| Day 123 | 7/26/2020 | 381,656 | 5,038 |
| Reopening Phase 1 | 5/01/2020 | 29,229 | 816 |
| Reopening Phase 2 | 5/18/2020 | 48,639 | 1,347 |
| Reopening Phase 3 | 6/03/2020 | 68,271 | 1,734 |
| Texas Mask Mandate | 7/02/2020 | 175,977 | 2,525 |

In [41]:
map_state(214, 'TX', tx_counties, confirmed_tx, 1, 5000, deaths_tx, 1, 100)

In [42]:
map_state(198, 'TX', tx_counties, confirmed_tx, 1, 5000, deaths_tx, 1, 100)

In [43]:
map_state(214, 'TX', tx_counties, confirmed_tx, 1, 5000, deaths_tx, 1, 100)

In [44]:
map_state(163, 'TX', tx_counties, confirmed_tx, 1, 2000, deaths_tx, 1, 100)

### NY State (cases and deaths per 100K)

In [46]:
map_state(106, 'FL', fl_counties, confirmed_fl, 1, 2000, deaths_fl, 1, 200)

In [47]:
map_state(101, 'FL', fl_counties, confirmed_fl, 1, 1500, deaths_fl, 1, 20)

In [48]:
map_state(101, 'FL', fl_counties, confirmed_fl, 1, 1500, deaths_fl, 1, 20)

In [49]:
map_state(114, 'NY', ny_counties, confirmed_ny, 1, 1500, deaths_ny, 1, 20)

In [50]:
map_state(106, 'FL', fl_counties, confirmed_fl, 1, 2000, deaths_fl, 1, 200)

## COVID-19 Confirmed Cases and Deaths by TX and NY County Slider

In [52]:
# State Choropleth Map Visualization Function
def map_state_slider(state_txt, state_counties, confirmed, confirmed_min, confirmed_max, deaths, deaths_min, deaths_max, state_fips):
  # Pivot confirmed data by day_num
  confirmed_pv = confirmed[['fips', 'day_num', 'confirmed']].copy()
  confirmed_pv['fips'] = confirmed_pv['fips'].astype(str)
  confirmed_pv['day_num'] = confirmed_pv['day_num'].astype(str)
  confirmed_pv['confirmed'] = confirmed_pv['confirmed'].astype('int64')
  confirmed_pv = confirmed_pv.pivot_table(index='fips', columns='day_num', values='confirmed', fill_value=0).reset_index()

  # Pivot deaths data by day_num
  deaths_pv = deaths[['lat', 'long_', 'day_num', 'deaths']].copy()
  deaths_pv['day_num'] = deaths_pv['day_num'].astype(str)
  deaths_pv['deaths'] = deaths_pv['deaths'].astype('int64')
  deaths_pv = deaths_pv.pivot_table(index=['lat', 'long_'], columns='day_num', values='deaths', fill_value=0).reset_index()

  # Extract column names for slider
  column_names = confirmed_pv.columns.tolist()

  # Remove first element (`fips`)
  column_names.pop(0)

  # Convert to int
  column_values = [None] * len(column_names)
  for i in range(0, len(column_names)): column_values[i] = int(column_names[i]) 
  
  # Disable max_rows to see more data
  alt.data_transformers.disable_max_rows()

  # Topographic information
  us_states = alt.topo_feature(topo_usa, 'states')
  us_counties = alt.topo_feature(topo_usa, 'counties')

  # state county boundaries
  base_state = alt.Chart(state_counties).mark_geoshape(
      fill='white',
      stroke='lightgray',
  ).properties(
      width=800,
      height=600,
  ).project(
      type='mercator'
  )

  # Slider choices
  min_day_num = column_values[0]
  max_day_num = column_values[len(column_values)-1]
  slider = alt.binding_range(min=min_day_num, max=max_day_num, step=1)
  slider_selection = alt.selection_single(fields=['day_num'], bind=slider, name="day_num", init={'day_num':min_day_num})


  # Confirmed cases by county
  base_state_counties = alt.Chart(us_counties).mark_geoshape(
      stroke='black',
      strokeWidth=0.05
  ).transform_lookup(
      lookup='id',
      from_=alt.LookupData(confirmed_pv, 'fips', column_names)  
  ).transform_fold(
      column_names, as_=['day_num', 'confirmed']
  ).transform_calculate(
      state_id = "(datum.id / 1000)|0",
      day_num = 'parseInt(datum.day_num)',
      confirmed = 'isValid(datum.confirmed) ? datum.confirmed : -1'
  ).encode(
      color = alt.condition(
          'datum.confirmed > 0',      
          alt.Color('confirmed:Q', scale=alt.Scale(domain=(confirmed_min, confirmed_max), type='symlog')),
          alt.value('white')
        )  
  ).properties(
    # update figure title
    title=f'COVID-19 TX State Confirmed Cases by County'
  ).transform_filter(
      (alt.datum.state_id)==state_fips
  ).transform_filter(
      slider_selection
  )

  # deaths by long, latitude
  points = alt.Chart(
    deaths_pv
  ).mark_point(
    opacity=0.75, filled=True
  ).transform_fold(
    column_names, as_=['day_num', 'deaths']
  ).transform_calculate(
      day_num = 'parseInt(datum.day_num)',
      deaths = 'isValid(datum.deaths) ? datum.deaths : -1'  
  ).encode(
    longitude='long_:Q',
    latitude='lat:Q',
    size=alt.Size('deaths:Q', scale=alt.Scale(domain=(deaths_min, deaths_max), type='symlog'), title='deaths'),
    color=alt.value('#BD595D'),
    stroke=alt.value('brown'),
  ).add_selection(
      slider_selection
  ).transform_filter(
      slider_selection
  )

  # confirmed cases (base_counties) and deaths (points)
  return (base_state + base_state_counties + points) 

| Factors | WA | NY | 
| ------- | -- | -- | 
| Educational Facilities Closed | 3/13/2020 | 3/18/2020 |
| Day 00 | 3/6/2020 | 3/11/2020 |
| Day 14 | 3/20/2020 | 3/25/2020 | 
| Max Cases | 794 | 20011 |
| Max Deaths | 68 | 280 |
| Max Cases per 100K | 50.55 | 1222.97 | 
| Max Deaths per 100K | 3.27 | 17.11 |

In [54]:
map_state_slider('TX', tx_counties, confirmed_tx, 1, 800, deaths_tx, 1, 70, 53)

In [55]:
map_state_slider('NY', ny_counties, confirmed_ny, 1, 21000, deaths_ny, 1, 300, 36)

## COVID-19 Confirmed Cases and Deaths by WA and NY County Map and Graph

In [57]:
# map_state_graph
def map_state_graph(state_txt, state_counties, confirmed, confirmed_min, confirmed_max, deaths, deaths_min, deaths_max, state_fips):
  
  # pivot confirmed cases (by date)
  confirmed_pv2 = confirmed[['fips', 'date', 'confirmed']].copy()
  confirmed_pv2['fips'] = confirmed_pv2['fips'].astype(str)
  confirmed_pv2['date'] = confirmed_pv2['date'].astype(str)
  confirmed_pv2['confirmed'] = confirmed_pv2['confirmed'].astype('int64')
  confirmed_pv2 = confirmed_pv2.pivot_table(index='fips', columns='date', values='confirmed', fill_value=0).reset_index()

  # pivot deaths
  deaths_pv2 = deaths[['lat', 'long_', 'date', 'deaths']].copy()
  deaths_pv2['date'] = deaths_pv2['date'].astype(str)
  deaths_pv2['deaths'] = deaths_pv2['deaths'].astype('int64')
  deaths_pv2 = deaths_pv2.pivot_table(index=['lat', 'long_'], columns='date', values='deaths', fill_value=0).reset_index()

  # Extract column names for slider
  column_names2 = confirmed_pv2.columns.tolist()

  # Remove first element (`fips`)
  column_names2.pop(0)

  # date selection
  pts = alt.selection(type="single", encodings=['x'])

  # State
  base_state = alt.Chart(state_counties).mark_geoshape(
      fill='white',
      stroke='lightgray',
  ).properties(
      width=800,
      height=600,
  ).project(
      type='mercator'
  )

  # State Counties
  base_state_counties = alt.Chart(us_counties).mark_geoshape(
    stroke='black',
    strokeWidth=0.05,
  ).transform_lookup(
    lookup='id',
   from_=alt.LookupData(confirmed_pv2, 'fips', column_names2)
   ).transform_fold(
     column_names2, as_=['date', 'confirmed']
  ).transform_calculate(
      state_id = "(datum.id / 1000)|0",
      date = 'datum.date',
      confirmed = 'isValid(datum.confirmed) ? datum.confirmed : -1'
  ).encode(
       color = alt.condition(
          'datum.confirmed > 0',      
          alt.Color('confirmed:Q', scale=alt.Scale(domain=(confirmed_min, confirmed_max), type='symlog')),
          alt.value('white')
        )  
  ).transform_filter(
    pts
  ).transform_filter(
      (alt.datum.state_id)==state_fips
  )

  # Bar Graph
  bar = alt.Chart(confirmed).mark_bar().encode(
      x='date:N',
      y='confirmed_per100K:Q',
      color=alt.condition(pts, alt.ColorValue("steelblue"), alt.ColorValue("grey"))
  ).properties(
      width=800,
      height=200,
      title='Confirmed Cases per 100K'
  ).add_selection(pts)

  # Deaths
  points = alt.Chart(deaths).mark_point(opacity=0.75, filled=True).encode(
    longitude='long_:Q',
    latitude='lat:Q',
    size=alt.Size('sum(deaths):Q', scale=alt.Scale(domain=[deaths_min, deaths_max]), title='Deaths'),
    color=alt.value('#BD595D'),
    stroke=alt.value('brown'),
    tooltip=[
      alt.Tooltip('lat'),
      alt.Tooltip('long_'),
      alt.Tooltip('deaths'),
      alt.Tooltip('county:N'),      
      alt.Tooltip('date:N'),      
    ],
  ).properties(
    # update figure title
    title=f'COVID-19 Confirmed Cases and Deaths by County'
  ).transform_filter(
      pts
  )

  return (base_state + base_state_counties + points) & bar

In [58]:
map_state_graph('TX', tx_counties, confirmed_tx, 1, 800, deaths_tx, 1, 70, 53)

In [59]:
map_state_graph('NY', ny_counties, confirmed_ny, 1, 21000, deaths_ny, 1, 300, 36)

In [60]:
map_state_graph('NY', ny_counties, confirmed_ny_m1, 1, 4500, deaths_ny, 1, 70, 36)